### DISCUSSION TITLE
Textual help

### DISCUSSION THREAD

**Discussion by UntriexTv** [Textualize/textual#26] [Q&A]
Hello. I am working on a project for Jam, and i want to make a dashboard. I used rich, but i cant bind keys or use mouse in it so i tried textual. But in textual because of limited sources i cant find any way to update widgets when program is running (like Live function in rich). Is this feature even implemented yet?
![image](https://user-images.githubusercontent.com/48880229/125158696-0c5f6880-e173-11eb-9207-1985b66adafd.png)
Could you please help me?
thank you
*edit:
Also can i make new sidebars when program is running?

**Comment by lllama** [comment]
What happens when you update your Rich component? I want to say that if you change what you want to display then Textual will handle the updating. I.e. no need to explicitly update, just change your data and let Textual handle it.

That being said, the examples all seem to be static content at the moment, so 🤷🏻‍♂️

@willmcgugan could be worth adding current time to one of the placeholders in the examples so people can see something updating? Just a thought.

**Reply by UntriexTv** [reply]
Nothing happens (if i am doing it right).
I am using widget and my definition that returns randomly generated table.
![image](https://user-images.githubusercontent.com/48880229/125159891-74657d00-e17a-11eb-86b5-4fbab1973f2d.png)
Then i am trying to generate and show new table (replace the old table) and tried lot of choices.
RN i tried this
![image](https://user-images.githubusercontent.com/48880229/125159921-95c66900-e17a-11eb-9741-0dfed9f763a0.png)
I found out that this function runs every second so i want to use it to update widgets. But nothing works.

**Reply by willmcgugan** [reply]
@lllama The placeholders will change when you hover the mouse or click to focus. There's an updating clock on the header...

**Comment by willmcgugan** [comment]
You can do this, but you would be a super early adopter. If you are brave you could give something like this a try:

```python
class LayoutApp(App):

    async def on_load(self, event):
        await self.bind("q", "quit")
        await self.bind("b", "bell")

    async def action_bell(self):
        self.console.bell()

    async def get_layout(self) -> Layout:
        # Build your layout here
  
    async def on_startup(self, event:events.Startup) -> None:
        self.layout_widget = Static(self.get_layout())
        await self.dock(layout_widget)
        await self.set_interval(0.1, callback=get_layout)
  
    async def update_layout(self) -> None:
        self.layout_widget.update(self.get_layout())

LayoutApp.run()
```

I haven't tested that, so can't guarantee it will work. You may have to figure out the details from the code.

**Reply by UntriexTv** [reply]
I edited it so there are no errors, but still no updating:
```python
class LayoutApp(App):

    async def on_load(self, event):
        await self.bind("q", "quit")
        await self.bind("b", "bell")

    async def action_bell(self):
        self.console.bell()

    async def get_layout(self) -> Layout:
        # Build your layout here
  
    async def on_startup(self, event:events.Startup) -> None:
        self.layout_widget = Static(self.get_layout())
        await self.dock(layout_widget)
        await self.set_interval(0.1, callback=get_layout)
  
    async def update_layout(self) -> None:
        self.layout_widget.update(self.get_layout())

LayoutApp.run()
```

**Reply by willmcgugan** [reply]
Try `self.layout_widget.require_repaint()` in update_layout. But I'm afraid you're own your own if it doesn't work. It's too early for me to offer support when I'm still actively building it... Good luck tho.

**Reply by UntriexTv** [reply]
Managed to get it working
```python
class LayoutApp(App):
    async def on_load(self, event):
        await self.bind("q", "quit")
        await self.bind("b", "bell")

    async def action_bell(self):
        self.console.bell()

    async def get_layout(self):
        table = TestingWidget("")
        return table.run()

    async def on_startup(self, event: events.Startup) -> None:
        view = await self.push_view(DockView())
        self.layout_widget = Static(await self.get_layout())
        await view.dock(self.layout_widget)
        view.set_interval(0.1, callback=self.update_layout)

    async def update_layout(self) -> None:
        self.layout_widget.renderable = await self.get_layout()
        self.layout_widget.require_repaint()


LayoutApp.run()
```
Thank you so much for your help

**Reply by willmcgugan** [reply]
Aha! Sorry, the widget.update method is in my local branch. Nicely done.

**Reply by kfmfe04** [reply]
A quick update for anyone coming across this code:
  `self.layout_widget.refresh()`
appears to be the new replacement for:
  `self.layout_widget.require_repaint()`

And do:
  `on_mount()`
instead of:
  `on_startup()`

**Comment by UntriexTv** [comment]
I found another problem. I can not run my definition with key bindings in textual. Is there any way to do that?

**Reply by willmcgugan** [reply]
Not sure what you mean by "run my definition" ?

**Reply by UntriexTv** [reply]
def in myapp:
![image](https://user-images.githubusercontent.com/48880229/125168944-4ac34a80-e1a8-11eb-891c-6692539476c6.png)

In another word how to run function on keypress

**Reply by willmcgugan** [reply]
You can bind it in `on_load`:

```
await self.bind("l", "change_l('whatever')")
```

or you can define an on_key event handler:

```
async def on_key(self, event):
    # event.key contains the "l" or whatever.
```

You won't be able to print anything when the terminal is in application mode. Suggest you using logging. No need to return anything from the action....

**Reply by UntriexTv** [reply]
i binded it  on_load but it does not run the function
![image](https://user-images.githubusercontent.com/48880229/125169324-f28d4800-e1a9-11eb-966d-eab22d2b4c1b.png)
i will try second one

**Reply by UntriexTv** [reply]
![image](https://user-images.githubusercontent.com/48880229/125169729-cb377a80-e1ab-11eb-926f-f5cc11491f57.png)
i tried other thinks too (not only print)
And second one did not work too

**Reply by UntriexTv** [reply]
Ok i somehow get it to work. I downloaded newest source code from branch and replaced version downloaded from pip. 
`python
await self.bind("b", "change", "Testing")
`
`python
async def update_layout(self):
`
Probably "()" and "_" were the problem
